erts/emulator/beam/erl_bif_binary.c

```diff
@@ -875,22 +892,44 @@ static BFReturn bm_find_all_non_overlapping(BinaryFindContext *ctx, byte *haysta
     Sint *gs = bmd->goodshift;
     Sint *bs = bmd->badshift;
     byte *needle = bmd->x;
-    Sint i;
+    Sint i = -1; /* Use memchr on start and on every match */
     Sint j = state->pos;
     Uint m = state->m;
     Uint allocated = state->allocated;
     FindallData *out = state->out;
     register Uint reds = *reductions;
+    byte *pos_pointer;
+    Sint needle_last = blen - 1;
+    Sint mem_read;
 
-    while (j <= len - blen) {
-	if (--reds == 0) {
+    for(;;) {
+	if (i < 0) {
+	    mem_read = MIN(len - j, reds * MC_LOOP_FACTOR);
+	    ASSERT(mem_read > 0);
+	    pos_pointer = memchr(&haystack[j], needle[needle_last], mem_read);
+	    if (pos_pointer == NULL) {
+		j += mem_read;
+		reds -= mem_read / MC_LOOP_FACTOR;
+	    } else {
+		j = pos_pointer - haystack - needle_last;
+		reds -= (pos_pointer - haystack - needle_last - j) / MC_LOOP_FACTOR;
```

Comment:
Maybe I'm reading this wrong, but won't this always be equivalent to `reds -= (j - j) / MC_LOOP_FACTOR;`?  Or in other words: `reds -= 0;`?

Reply:
Is this perhaps supposed to be something like this?
```c
reds -= (pos_pointer - &haystack[j] - needle_last) / MC_LOOP_FACTOR;
j = pos_pointer - haystack - needle_last;
```